docs/reconfigurator.adoc

```diff
@@ -0,0 +1,183 @@
+:showtitle:
+:numbered:
+:toc: left
+
+= Reconfigurator
+
+== Introduction
+
+**Reconfigurator** is a system within Nexus for carrying out all kinds of system changes in a controlled way.  Examples of what Reconfigurator can do today or that we plan to extend it to do in the future:
+
+* add a new sled to a system (deploying any components it needs to be able to run customer workloads or participate in the control plane)
+* expunge a sled from a system (i.e., respond to the permanent physical removal of a sled)
+* deploy a new instance of a component (e.g., Nexus) to scale it out or replace one that's failed
+* remove an instance of a component (e.g., Nexus) to scale it down or as part of moving it to another sled
+* recover a sled whose host software is arbitrarily broken (similar to mupdate)
+* update the host OS, service processor, control plane components, and other software on the sled
+
+// TODO an omdb demo here would be useful
+
+Let's walk through how someone might add a new sled to an existing system using Reconfigurator.
+
+1. The user makes an API request to add the sled.footnote:[There is a bit more to this flow.  There's an API request to list sleds that are physically present but not part of the system.  The user is expected to compare that list against what they expect and then make an API request to add the specific sled they expect to be there (by serial number).]  This request succeeds immediately but has not done most of the work yet.
+2. The user makes an API request to invoke the **planner** (one of two main parts of Reconfigurator).  The planner generates a new **blueprint**.
+3. Optional: the user has a chance to compare the new blueprint against the existing one.  This lets them see what the system is going to do before committing to it.
+4. The user makes an API request to make the new blueprint the current system **target blueprint**.
+5. Reconfigurator does the rest: it starts **executing** the blueprint.  This means trying to make reality match the blueprint.  This process continues in a loop basically forever.footnote:[The process does not stop once reality matches the blueprint because reality can change after that point and the system may need to take action again.]
+
+NOTE: Currently, these "user"-driven steps are carried out using `omdb`, which means that in production systems they can really only be done by an Oxide employee with access via the technician port.  The long-term goal is to automate all of this so that planning and execution both happen continuously and automatically.
+
+You might notice that the only step that has anything to do with adding a sled is the first one.  This is where a user communicated a change to the **intended state** of the world.  Everything Reconfigurator does follows a pattern just like this:
+
+1. A user changes the intended state of the world.
```

Comment:
Potentially there should be a second step where the inventory is reloaded to reflect the latest state of the world. The whole system acts as a feedback loop around planning input (mostly inventory, db state, and policy).

Reply:
In dba4f803c I added a step _before_ this just to say that Reconfigurator is always updating its view of the world.

I'm not sure if it's clearer or more confusing, to be honest.  In practice, an inventory collection might not happen immediately before, during, or after this sequence.  And I feel like inventories are less relevant than it seems like they would be (since they're always out of date and potentially incomplete anyway).  We don't use them most of the time during planning or execution -- instead we propagate the full intended state everywhere, every time (to sled agents, DNS servers, etc.).

It's a small difference either way so I just kept this in.